includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -1880,6 +1905,109 @@ protected function get_account_branding_icon( $default_value = '' ): string {
+	 *
+	 * @return array|void A successful response in case the session processing order was paid, null if none.
+	 */
+	protected function check_against_session_processing_order( WC_Order $current_order ) {
+		$session_order_id = $this->get_session_processing_order();
+		if ( null === $session_order_id ) {
+			return;
+		}
+
+		$session_order = wc_get_order( $session_order_id );
+		if ( ! is_a( $session_order, 'WC_Order' ) ) {
+			return;
+		}
+
+		if ( $current_order->get_cart_hash() !== $session_order->get_cart_hash() ) {
+			return;
+		}
+
+		if ( ! $session_order->has_status( wc_get_is_paid_statuses() ) ) {
+			return;
+		}
+
+		$session_order->add_order_note(
+			sprintf(
+				/* translators: order ID integer number */
+				__( 'WooCommerce Payments: detected and deleted order ID %d, which has duplicate cart content with this order.', 'woocommerce-payments' ),
+				$current_order->get_id()
+			)
+		);
+		$current_order->delete();
+
+		$this->remove_session_processing_order( $session_order_id );
+
+		$return_url = $this->get_return_url( $session_order );
+		$return_url = add_query_arg( self::FLAG_PREVIOUS_ORDER_PAID, 'yes', $return_url );
+
+		return [
+			'result'                            => 'success',
+			'redirect'                          => $return_url,
+			'wcpay_upe_paid_for_previous_order' => 'yes', // This flag is needed for UPE flow.
+		];
+	}
+
+	/**
+	 * Update the processing order ID for the current session.
+	 *
+	 * @param  int $order_id Order ID.
+	 *
+	 * @return void
+	 */
+	protected function maybe_update_session_processing_order( int $order_id ) {
+		if ( WC()->session ) {
+			WC()->session->set( self::SESSION_KEY_PROCESSING_ORDER, $order_id );
+		}
+	}
+
+	/**
+	 * Remove the provided order ID from the current session if it matches with the ID in the session.
+	 *
+	 * @param  int $order_id Order ID to remove from the session.
+	 *
+	 * @return void
+	 */
+	protected function remove_session_processing_order( int $order_id ) {
+		$current_session_id = $this->get_session_processing_order();
+		if ( $order_id === $current_session_id && WC()->session ) {
+			WC()->session->set( self::SESSION_KEY_PROCESSING_ORDER, null );
+		}
+	}
+
+	/**
+	 * Get the processing order ID for the current session.
+	 *
+	 * @return integer|null Order ID. Null if the value is not set.
+	 */
+	protected function get_session_processing_order() {
+		$session = WC()->session;
+		if ( null === $session ) {
+			return null;
+		}
+
+		$val = $session->get( self::SESSION_KEY_PROCESSING_ORDER );
+		return null === $val ? null : absint( $val );
+	}
+
+	/**
+	 * Action to remove the order ID when customers reach its order-received page.
+	 *
+	 * @return void
+	 */
+	public function clear_session_processing_order_after_landing_order_received_page() {
+		global $wp;
+
+		if ( is_order_received_page() && isset( $wp->query_vars['order-received'] ) ) {
+			$order_id = absint( $wp->query_vars['order-received'] );
```

Comment:
A small note, nothing to fix: I was wondering what happens if the `woocommerce_order_number` filter is used to change the order number, and if that might break here, but the URL still includes the order ID instead of number.

Reply:
Good point. Marking as resolved.